mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,662 @@
+        )
+        obs = switch_leading_axes(obs)  # (B, T) -> (T, B)
+        done = switch_leading_axes(done)  # (B, T) -> (T, B)
+        obs_done = (obs, done)
+
+        return hidden_state, obs_done
+
+    # TRAIN LEVEL 3
+    def q_loss_fn(
+        q_online_params: FrozenVariableDict, obs: Array, done: Array, action: Array, target: Array
+    ) -> Tuple[Array, Metrics]:
+        """The portion of the calculation to grad, namely online apply and mse with target."""
+
+        hidden_state, obs_done = prep_inputs_to_scannedrnn(
+            obs, done
+        )  # axes switched here to scan over time
+
+        _, q_online = q_net.apply(
+            q_online_params, hidden_state, obs_done, method="get_q_values"
+        )  # get online q values of all actions
+        q_online = switch_leading_axes(q_online)  # (T, B, ...) -> (B, T, ...)
+        q_online = jnp.squeeze(
+            jnp.take_along_axis(q_online, action[..., jnp.newaxis], axis=-1), axis=-1
+        )  # get the q values of the taken actions and remove extra dim
+        q_loss = jnp.mean((q_online - target) ** 2)  # mse
+
+        # pack metrics for logging
+        loss_info = {
+            "q_loss": q_loss,
+            "mean_q": jnp.mean(q_online),
+            "max_q_error": jnp.max(jnp.abs(q_online - target) ** 2),
+            "min_q_error": jnp.min(jnp.abs(q_online - target) ** 2),
+            "mean_target": jnp.mean(target),
+        }
+
+        return q_loss, loss_info
+
+    # Standardise the update function inputs for a cond
+    def hard_update(
+        next_online_params: FrozenVariableDict, target_params: FrozenVariableDict, t_train: int
+    ) -> FrozenVariableDict:
+        next_target_params = optax.periodic_update(
+            next_online_params, target_params, t_train, cfg.system.update_period
+        )
+        return next_target_params
+
+    def soft_update(
+        next_online_params: FrozenVariableDict, target_params: FrozenVariableDict, t_train: int
+    ) -> FrozenVariableDict:
+        next_target_params = optax.incremental_update(
+            next_online_params, target_params, cfg.system.tau
+        )
+        return next_target_params
+
+    # TRAIN LEVEL 2
+    def update_q(
+        params: DDQNParams, opt_states: optax.OptState, data: Transition, t_train: int
+    ) -> Tuple[DDQNParams, optax.OptState, Metrics]:
+        """Update the Q parameters."""
+
+        # Get the data associated with obs
+        data_first = jax.tree_map(lambda x: x[:, :-1, ...], data)
+        # Get the data associated with next_obs
+        data_next = jax.tree_map(lambda x: x[:, 1:, ...], data)
+
+        first_reward = data_first.reward
+        next_done = data_next.done
+
+        # Scan over each sample and discard first timestep
+        hidden_state, obs_done = prep_inputs_to_scannedrnn(data.obs, data.done)
+
+        _, next_online_greedy_dist = q_net.apply(
+            params.online, hidden_state, obs_done
+        )  # eps defaults to 0
```

Comment:
I feel like this is 90% of my suggestions lol
```suggestion
        # eps defaults to 0
        _, next_online_greedy_dist = q_net.apply(
            params.online, hidden_state, obs_done
        ) 
```